sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -83,23 +83,29 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
   return Program;
 }
 
-RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
-                                                    const context &Context) {
+RT::PiProgram
+ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M, const context &Context,
+                                      const string_class &KernelName) {
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
-  std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
+  std::map<OSModuleHandle, std::vector<RT::PiProgram>> &CachedPrograms =
       Ctx->getCachedPrograms();
   auto It = CachedPrograms.find(M);
-  if (It != CachedPrograms.end())
-    return It->second;
+  if (It != CachedPrograms.end()) {
+    for (auto Prg : It->second) {
+      if (programContainsKernel(Prg, KernelName))
+        return Prg;
+    }
+  }
 
   DeviceImage *Img = nullptr;
   using PiProgramT = remove_pointer_t<RT::PiProgram>;
   unique_ptr_class<PiProgramT, decltype(RT::piProgramRelease)> ProgramManaged(
-      loadProgram(M, Context, &Img), RT::piProgramRelease);
+      loadProgram(M, Context, KernelName, &Img), RT::piProgramRelease);
```

Comment:
here and in some other places, kernel name should not be added to internal program management APIs. Any kernel resolution (finding device binary or Program for given kernel) should be separate from program management itself and done up the call stack.

Reply:
@kbobrovs At the moment device images (and their descriptors) are encapsulated within the program manager, so it makes sense to me to have this kernel resolution logic as a part of the program manager as well, although considering your comments regarding the PI interface, it definitely should be separate from program creation itself. But from what I inferred from your comment, you're saying that this logic should be decoupled from the program manager class altogether. Is that correct?

Reply:
> it makes sense to me to have this kernel resolution logic as a part of the program manager as well

I agree

> you're saying that this logic should be decoupled from the program manager class altogether. Is that correct

not quite. I'm just saying that kernel resolution should be intermixed with program loading. I consider adding  kernel name parameter to most of internal program management functions to be such intermixing. But program manager can continue to be responsible for both.